docs/transpile/custom-transpiler-pass.ipynb

```diff
@@ -0,0 +1,352 @@
+    "## Transpiler Passes\n",
+    "\n",
+    "Transpiler passes are classified into one of two classes: Analysis passes and Transformation passes. Passes in general work with the Directed Acyclic Graph (DAG) **TODO: insert link here** and the `property_set`, a dictionary-like object for storing properties determined by analysis passes. Analysis passes work with the DAG, but are not allowed to modify it. This contrasts with transformation passes, which do modify the DAG, and can read (but not write to) `property_set`. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Internals\n",
+    "\n",
+    "**TODO: put something here, but I do not know what**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create a `PauliTwirl` transpiler pass\n",
+    "\n",
+    "As an example we construct a transpiler pass that adds Pauli twirls. Pauli twirling is an error suppression strategy that randomizes how the qubits experience noisy channels, which we assume to be two-qubit gates in this example (because they are much more error-prone than single-qubit gates). However, the Pauli twirls do not affect the actual operation of the two-qubit gate. They are chosen such that those applied *before* the two-qubit gate (to the left) are countered by those applied *after* the two-qubit gate (to the right). In this sense, the two-qubit operations are identical but the way they are performed is different. One benefit of Pauli twirling is that it turns coherent errors into stochastic errors, which can be improved by averaging more.\n",
+    "\n",
+    "Transpiler passes act on the DAG (directed acyclic graph **TODO: include link**), so the important method to override is `.run()`, which takes the DAG as the input. Here we will also add the initialization of pairs of Paulis that preserve the operation of each two-qubit gate. This is done with the helper method `build_twirl_set` that goes through each two-qubit Pauli (as obtained from `pauli_basis(2)`) and finding the other Pauli that preserves the operation. \n",
+    "\n",
+    "From the DAG we can collect runs/nodes (**TODO: what really is the difference between these**), which provides the set of every gate with a given name in the DAG. **TODO: what other ways can you search, like `.bfs_successors` and such? Does this belong here or in the DAG doc?**. Here, we wish to replace each node, representing an instruction, with a subcircuit which we build as a mini DAG. The mini DAG has a two-qubit quantum register added to it, and operations are added here with `apply_operation_back`, which places the `Instruction` on the output of the mini DAG (whereas `apply_operation_front` would place it on the input of the mini DAG). The node is then substituted by the mini DAG with `substitute_node_with_dag`, and the process continues over each instance of `CXGate` and `ECRGate` in the DAG (corresponding to the two-qubit basis gates on IBM backends)."
```

Comment:
A run is just a sequence of nodes, the best example is https://qiskit.org/documentation/stubs/qiskit.dagcircuit.DAGCircuit.html#qiskit.dagcircuit.DAGCircuit.collect_1q_runs which returns a sequence of all the 1q gates that run uninterrupted in a sequence on a qubit.

The dagcircuit has many methods for searching and traversing a graph (and there are even more available in graph theory more generally), the most common method used is probably `topological_op_nodes`: https://qiskit.org/documentation/stubs/qiskit.dagcircuit.DAGCircuit.html#qiskit.dagcircuit.DAGCircuit.topological_op_nodes to get the gates in a dependency ordering. Things like `bfs_successors` are used primarily to determine how gates interact with subsequent operations on a DAG.

Reply:
Thanks!